libraries/customisations/src/main/kotlin/com/bumble/appyx/utils/customisations/NodeCustomisationDirectoryImpl.kt

```diff
@@ -32,21 +37,42 @@ open class NodeCustomisationDirectoryImpl(
         map[key] = value
     }
 
-    override fun <T : Any> getSubDirectory(key: KClass<T>): NodeCustomisationDirectory?=
-        map[key] as? NodeCustomisationDirectory
+    override fun <T : Any> putLazySubDirectory(
+        key: KClass<T>,
+        valueProvider: () -> NodeCustomisationDirectory
+    ) {
+        map[key] = lazy(valueProvider)
+    }
+
+    override fun <T : Any> getSubDirectory(key: KClass<T>): NodeCustomisationDirectory? =
+        map.lazyGet(key)
 
     override fun <T : Any> getSubDirectoryOrSelf(key: KClass<T>): NodeCustomisationDirectory {
         val subDir = map.keys.firstOrNull {
             it is KClass<*> && it.java.isAssignableFrom(key.java)
         }
 
-        return map[subDir] as? NodeCustomisationDirectory ?: this
+        return map.lazyGet(subDir) ?: this
     }
 
     operator fun KClass<*>.invoke(block: NodeCustomisationDirectoryImpl.() -> Unit) {
         if (map.containsKey(this)) {
             // TODO warning for accidental override?
         }
-        map[this] = NodeCustomisationDirectoryImpl(this@NodeCustomisationDirectoryImpl).apply(block)
+        map[this] = lazy {
+            NodeCustomisationDirectoryImpl(this@NodeCustomisationDirectoryImpl).apply(block)
+        }
     }
+
+    /**
+     * Gets a value from the map, if the value is lazy the lazy value will be initialized.
+     */
+    private fun <T: Any> MutableMap<Any, Any>.lazyGet(key: Any?): T? =
+        get(key).let { customisation ->
+            if (customisation is Lazy<*>) {
```

Comment:
I would say it is kinda of dangerous (1 step more dangerous than it is right now).

Maybe
```kotlin
val map : MutableMap<Any, Holder> = hashMapOf()

sealed class Holder {
  class Initialized(value)
  class Lazy(lamda)
}
```

Reply:
Yeah... I could wrap it with a sealed class instead? Thoughts?

I.e.

Sealed class Item {
     Data class Lazy(Val lazyitem: Any)
     Data class NotLazy(Val item: Any)
}

I also thought about having two different maps, but it got a bit ugly

Reply:
done @CherryPerry